File: src/main/java/work/xeltica/craft/core/models/PlayerRecord.java
package work.xeltica.craft.core.models;

import org.bukkit.Location;
import org.bukkit.configuration.ConfigurationSection;
import org.bukkit.util.Vector;
import work.xeltica.craft.core.api.Config;
import work.xeltica.craft.core.stores.PlayerStore;

import java.io.IOException;
import java.util.List;
import java.util.UUID;

/**
 * プレイヤーごとのデータを読み書きするインターフェイスです。
 * @author Xeltica
 */
public class PlayerRecord {
    public PlayerRecord(Config conf, ConfigurationSection section, UUID playerId) {
        this.conf = conf;
        this.section = section;
        this.playerId = playerId;
    }

    public void set(PlayerDataKey key, Object value) {
        if (value != null && value.equals(get(key))) return;
        if (get(key) == null && value == null) return;
        section.set(key.getPhysicalKey(), value);
        PlayerStore.getInstance().setChanged(true);
    }

    @Deprecated(since = "Use set(PlayerDataKey, Object) instead.")
    public void set(PlayerDataKey key, Object value, boolean save) {
        set(key, value);
    }

    public boolean has(PlayerDataKey key) {
        return section.contains(key.getPhysicalKey());
    }

    public Object get(PlayerDataKey key) {
        return section.get(key.getPhysicalKey());
    }

    public Object get(PlayerDataKey key, Object defaultValue) {
        return section.get(key.getPhysicalKey(), defaultValue);
    }

    public String getString(PlayerDataKey key) {
        return section.getString(key.getPhysicalKey());
    }

    public String getString(PlayerDataKey key, String defaultValue) {
        return section.getString(key.getPhysicalKey(), defaultValue);
    }

    public boolean isString(PlayerDataKey key) {
        return section.isString(key.getPhysicalKey());
    }

    public List<String> getStringList(PlayerDataKey key) {
        return section.getStringList(key.getPhysicalKey());
    }

    public boolean isList(PlayerDataKey key) {
        return section.isList(key.getPhysicalKey());
    }

    public int getInt(PlayerDataKey key) {
        return section.getInt(key.getPhysicalKey());
    }

    public int getInt(PlayerDataKey key, int defaultValue) {
        return section.getInt(key.getPhysicalKey(), defaultValue);
    }

    public boolean isInt(PlayerDataKey key) {
        return section.isInt(key.getPhysicalKey());
    }

    public boolean getBoolean(PlayerDataKey key) {
        return section.getBoolean(key.getPhysicalKey());
    }

    public boolean getBoolean(PlayerDataKey key, boolean defaultValue) {
        return section.getBoolean(key.getPhysicalKey(), defaultValue);
    }

    public boolean isBoolean(PlayerDataKey key) {
        return section.isBoolean(key.getPhysicalKey());
    }

    public double getDouble(PlayerDataKey key) {
        return section.getDouble(key.getPhysicalKey());
    }

    public double getDouble(PlayerDataKey key, double defaultValue) {
        return section.getDouble(key.getPhysicalKey(), defaultValue);
    }

    public boolean isDouble(PlayerDataKey key) {
        return section.isDouble(key.getPhysicalKey());
    }

    public long getLong(PlayerDataKey key) {
        return section.getLong(key.getPhysicalKey());
    }

    public long getLong(PlayerDataKey key, long defaultValue) {
        return section.getLong(key.getPhysicalKey(), defaultValue);
    }

    public boolean isLong(PlayerDataKey key) {
        return section.isLong(key.getPhysicalKey());
    }

    public Vector getVector(PlayerDataKey key) {
        return section.getVector(key.getPhysicalKey());
    }

    public Vector getVector(PlayerDataKey key, Vector defaultValue) {
        return section.getVector(key.getPhysicalKey(), defaultValue);
    }

    public boolean isVector(PlayerDataKey key) {
        return section.isVector(key.getPhysicalKey());
    }

    public Location getLocation(PlayerDataKey key) {
        return section.getLocation(key.getPhysicalKey());
    }

    public Location getLocation(PlayerDataKey key, Location defaultValue) {
        return section.getLocation(key.getPhysicalKey(), defaultValue);
    }

    public boolean isLocation(PlayerDataKey key) {
        return section.isLocation(key.getPhysicalKey());
    }

    public void delete(PlayerDataKey key) {
        set(key, null);
    }

    @Deprecated(since = "Use set(PlayerDataKey, Object) instead.")
    public void delete(PlayerDataKey key, boolean save) {
        set(key, null, save);
    }

    private void save() {
        try {
            conf.save();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private final Config conf;
    private final ConfigurationSection section;

    private final UUID playerId;

    public UUID getPlayerId() {
        return this.playerId;
    }
}

File: src/main/java/work/xeltica/craft/core/modules/player/PlayerRecord.kt
package work.xeltica.craft.core.modules.player

import org.bukkit.Location
import org.bukkit.configuration.ConfigurationSection
import org.bukkit.util.Vector
import work.xeltica.craft.core.utils.Config
import java.io.IOException
import java.util.UUID

class PlayerRecord(private val config: Config, private val section: ConfigurationSection, val playerId: UUID) {
    operator fun set(key: PlayerDataKey, value: Any?) {
        if (value == get(key)) return
        section.set(key.physicalKey, value)
    }

    @JvmOverloads
    operator fun get(key: PlayerDataKey, defaultValue: Any? = null): Any? {
        return section.get(key.physicalKey, defaultValue)
    }

    fun has(key: PlayerDataKey): Boolean {
        return section.contains(key.physicalKey)
    }

    @JvmOverloads
    fun getString(key: PlayerDataKey, defaultValue: String? = null): String? {
        return section.getString(key.physicalKey, defaultValue)
    }

    fun isString(key: PlayerDataKey): Boolean {
        return section.isString(key.physicalKey)
    }

    fun getStringList(key: PlayerDataKey): MutableList<String> {
        return section.getStringList(key.physicalKey)
    }

    fun isList(key: PlayerDataKey): Boolean {
        return section.isList(key.physicalKey)
    }

    @JvmOverloads
    fun getInt(key: PlayerDataKey, defaultValue: Int = 0): Int {
        return section.getInt(key.physicalKey, defaultValue)
    }

    fun isInt(key: PlayerDataKey): Boolean {
        return section.isInt(key.physicalKey)
    }

    @JvmOverloads
    fun getBoolean(key: PlayerDataKey, defaultValue: Boolean = false): Boolean {
        return section.getBoolean(key.physicalKey, defaultValue)
    }

    fun isBoolean(key: PlayerDataKey): Boolean {
        return section.isBoolean(key.physicalKey)
    }

    @JvmOverloads
    fun getDouble(key: PlayerDataKey, defaultValue: Double = 0.0): Double {
        return section.getDouble(key.physicalKey, defaultValue)
    }

    fun isDouble(key: PlayerDataKey): Boolean {
        return section.isDouble(key.physicalKey)
    }

    @JvmOverloads
    fun getLong(key: PlayerDataKey, defaultValue: Long = 1): Long {
        return section.getLong(key.physicalKey, defaultValue)
    }

    fun isLong(key: PlayerDataKey): Boolean {
        return section.isLong(key.physicalKey)
    }

    @JvmOverloads
    fun getVector(key: PlayerDataKey, defaultValue: Vector? = null): Vector? {
        return section.getVector(key.physicalKey, defaultValue)
    }

    fun isVector(key: PlayerDataKey): Boolean {
        return section.isVector(key.physicalKey)
    }

    @JvmOverloads
    fun getLocation(key: PlayerDataKey, defaultValue: Location? = null): Location? {
        return section.getLocation(key.physicalKey, defaultValue)
    }

    fun isLocation(key: PlayerDataKey): Boolean {
        return section.isLocation(key.physicalKey)
    }

    fun delete(key: PlayerDataKey) {
        set(key, null)
    }

    fun save() {
        try {
            config.save()
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }
}